File: go/QnAMaker/rest-based-quickstart/quickstart.go
package main

// <snippet_dependencies>
import (
    "bytes"
    "encoding/json"
	"fmt"
	//"os"
    "io/ioutil"
    "net/http"
    "strconv"
	"time"
	"strings"
)
// </snippet_dependencies>

/* This sample for the Azure Cognitive Services QnA Maker API shows how to:
 * - Create an knowledge base.
 * - Get Operation status.
 * - Publish knowledge base - required to query knowledge base.
 * - Get endpoint key - required to query knowledge base.
 * - Query knowledge base.
 */

/*
 * Configure the local environment:
 *   Set the following environment variables on your local machine using the
 *   appropriate method for your preferred shell (Bash, PowerShell, Command Prompt, etc.).
 *
 *   QNAMAKER_AUTHORING_KEY
 *   QNAMAKER_AUTHORING_ENDPOINT
 *   QNAMAKER_RESOURCE_NAME
 *
 * If the environment variable is created after the application is launched in a console or with Visual
 * Studio, the shell (or Visual Studio) needs to be closed and reloaded for the environment variable to take effect.
 */

 // <snippet_response_structure>
 /// Represents the HTTP response returned by an HTTP request.
 type Response struct {
	Headers	map[string][]string
	Status  string
    Body	string
}
// </snippet_response_structure>

// <snippet_main>
func main() {

	var authoring_key string = os.Getenv("QNAMAKER_AUTHORING_KEY")

	// example - "MY-RESOURCE_NAME"
	var resource_name string =  os.Getenv("QNAMAKER_RESOURCE_NAME")

	// example: "https://MY-RESOURCE_NAME.cognitiveservices.azure.com/"
	var authoring_endpoint string = os.Getenv("QNAMAKER_AUTHORING_ENDPOINT")

	var service string = "qnamaker/v4.0"

	// 1) Create knowledge base - get knowledge base ID
	var knowledgeBaseID string = create_knowledgebase(service, authoring_key, authoring_endpoint)
	fmt.Println("knowledgeBaseID = " + knowledgeBaseID)

	// 2) Publish knowledge base
	publish(service, authoring_key, authoring_endpoint, knowledgeBaseID)

	// 3) Get published knowledge base's endpoint key
	var endpointKey string = getPublishedEndpoint(service, authoring_key, authoring_endpoint)
	fmt.Println("endpointKey = " + endpointKey)

	// 4) Query published knowledge base
	queryEndpoint(resource_name, endpointKey, knowledgeBaseID)
}
// </snippet_main>

// <snippet_get>
func get(key string, uri string) Response {

    req, _ := http.NewRequest("GET", uri, nil)
    req.Header.Add("Ocp-Apim-Subscription-Key", key)
    client := &http.Client{}
    response, err := client.Do(req)
    if err != nil {
        panic(err)
	}

    defer response.Body.Close()
    body, _ := ioutil.ReadAll(response.Body)
    return Response {response.Header, response.Status, string(body)}
}
// </snippet_get>

// <snippet_post>
// authoring param determines which authorization header to pass
func post(key string, uri string, content string, authoring bool) Response {

	var postContent = bytes.NewBuffer([]byte(content))
	var postContentLength = strconv.Itoa(len(content))
	var authorizationName string = ""
	var authorizationValue string =  ""

    req, _ := http.NewRequest("POST", uri, postContent)

	if authoring {
		authorizationName = "Ocp-Apim-Subscription-Key"
		authorizationValue = key
	} else {
		authorizationName = "Authorization"
		authorizationValue =  "EndpointKey " + key
	}

	req.Header.Add(authorizationName, authorizationValue)

	if(len(content)>0){
		req.Header.Add("Content-Type", "application/json")
		req.Header.Add("Content-Length", postContentLength)
	}

    client := &http.Client{}
    response, err := client.Do(req)
    if err != nil {
        panic(err)
	}

    defer response.Body.Close()
    body, _ := ioutil.ReadAll(response.Body)

    return Response {response.Header, response.Status, string(body)}
}
// </snippet_post>

// <snippet_check_status>
func check_status(authoring_key string, uri string) (string, string) {

	result := get(authoring_key, uri)

    if retry, success := result.Headers["Retry-After"]; success {
        return retry[0], result.Body
    } else {
        // If the response headers did not include a Retry-After value, default to 30 seconds.
        return "30", result.Body
    }
}
// </snippet_check_status>

// <snippet_create_knowledgebase>
func create_knowledgebase(service string, authoring_key string, authoring_endpoint string)(string){

	var authoring = true
	var uri string = authoring_endpoint + service + "/knowledgebases/create"

	var knowledgeBaseID string = ""

	var knowledgeBaseDef string = `{
		'name': 'QnA Maker FAQ',
		'qnaList': [
		  {
			'id': 0,
			'answer': 'You can use our REST APIs to manage your knowledge base. See here for details: https://westus.dev.cognitive.microsoft.com/docs/services/58994a073d9e04097c7ba6fe/operations/58994a073d9e041ad42d9baa',
			'source': 'Custom Editorial',
			'questions': [
			  'How do I programmatically update my knowledge base?'
			],
			'metadata': [
			  {
				'name': 'category',
				'value': 'api'
			  }
			]
		  }
		],
		'urls': [
		  'https://docs.microsoft.com/en-in/azure/cognitive-services/qnamaker/faqs'
		],
		'files': []
	  }`;

	result := post(authoring_key, uri, knowledgeBaseDef, authoring)

	operationIds := result.Headers["Location"]

	var done bool = false

	// wait for creation to complete
	for done == false {

        uri_checkstatus := authoring_endpoint + service + operationIds[0]
		wait, httpResponseBody := check_status(authoring_key, uri_checkstatus)

		// get property value from returned body's JSON
        var status_obj map[string]interface{}
        json.Unmarshal([]byte(httpResponseBody), &status_obj)
        state := status_obj["operationState"]

        // If the operation isn't finished, wait and query again.
        if state == "Running" || state == "NotStarted" {

            fmt.Printf ("Waiting " + wait + " seconds...")
            sec, _ := strconv.Atoi(wait)
            time.Sleep (time.Duration(sec) * time.Second)

        } else {
			done = true
			var knowledgeBaseResourceLocation = status_obj["resourceLocation"].(string)
			knowledgeBaseID = strings.ReplaceAll(knowledgeBaseResourceLocation, "/knowledgebases/", "")
        }
	}

	return knowledgeBaseID

}
// </snippet_create_knowledgebase>

// <snippet_publish>
func publish(service string, authoring_key string, authoring_endpoint string, knowledgeBaseId string){

	var authoring = true
	var uri string = authoring_endpoint + service + "/knowledgebases/" + knowledgeBaseId

	result := post(authoring_key, uri, "", authoring)

	// no data returned so check HTTP status 204
	fmt.Println("publish status code should be 204  = " + result.Status)
}
// </snippet_publish>

// <snippet_get_published_endpoint_key>
func getPublishedEndpoint(service string, authoring_key string, authoring_endpoint string)(string){

	var uri string = authoring_endpoint + service + "/endpointkeys"

	result := get(authoring_key, uri)

	// get property value from returned body's JSON
	var status_obj map[string]interface{}
	json.Unmarshal([]byte(result.Body), &status_obj)
	var endpointKey = status_obj["primaryEndpointKey"].(string)

	return endpointKey
}
// </snippet_get_published_endpoint_key>

// <snippet_query>
// query runtime instead of authoring
func queryEndpoint(resourceName string, endpoint_key string, knowledgeBaseId string){

	var authoring = false
	var uri string = "https://" + resourceName + ".azurewebsites.net/qnamaker/knowledgebases/" + knowledgeBaseId + "/generateAnswer";

	var question string = `{'question': 'Is the QnA Maker Service free?','top': 3}`;

	result := post(endpoint_key, uri, question, authoring)

	fmt.Println("result body = " + result.Body)
}
// </snippet_query>

File: java/qnamaker/sdk/quickstart.java
// <dependencies>
/* Download the following files.
 * - https://repo1.maven.org/maven2/com/microsoft/azure/cognitiveservices/azure-cognitiveservices-qnamaker/1.0.0-beta.1/azure-cognitiveservices-qnamaker-1.0.0-beta.1.jar
 * - https://repo1.maven.org/maven2/com/microsoft/azure/cognitiveservices/azure-cognitiveservices-qnamaker/1.0.0-beta.1/azure-cognitiveservices-qnamaker-1.0.0-beta.1.pom
 * Move the downloaded .jar file to a folder named "lib" directly under the current folder.
 * Rename the downloaded file to pom.xml.
 * At the command line, run
 * mvn dependency:copy-dependencies
 * This will download the .jar files depended on by azure-cognitiveservices-qnamaker-1.0.0-beta.1.jar to the folder "target/dependency" under the current folder. Move these .jar files to the "lib" folder as well.
 */
import com.microsoft.azure.cognitiveservices.knowledge.qnamaker.*;
import com.microsoft.azure.cognitiveservices.knowledge.qnamaker.models.*;

import java.io.*;
import java.lang.Object.*;
import java.time.format.DateTimeFormatter;  
import java.time.LocalDateTime; 
import java.util.*;
import java.net.*;
// </dependencies>


/* This sample does the following tasks.
 * Create a knowledge base.
 * Publish a knowledge base.
 * Download a knowledge base.
 * Query a knowledge base.
 * Delete a knowledge base.
*/

/* To compile and run, enter the following at a command prompt:
 * javac Quickstart.java -cp .;lib\*
 * java -cp .;lib\* Quickstart
 * This presumes your libraries are stored in a folder named "lib"
 * directly under the current folder. If not, please adjust the
 * -classpath/-cp value accordingly.
 */

public class Quickstart {
	// <resourceKeys>
	/* Configure the local environment:
	* Set the following environment variables on your local machine using the
	* appropriate method for your preferred shell (Bash, PowerShell, Command
	* Prompt, etc.).
	*
	* QNA_MAKER_SUBSCRIPTION_KEY
	* QNA_MAKER_ENDPOINT
	* QNA_MAKER_RUNTIME_ENDPOINT
	*
	* If the environment variable is created after the application is launched in a console or with Visual
	* Studio, the shell (or Visual Studio) needs to be closed and reloaded to take the environment variable into account.
	*/
    private static String authoring_key = System.getenv("QNA_MAKER_SUBSCRIPTION_KEY");
	private static String authoring_endpoint = System.getenv("QNA_MAKER_ENDPOINT");
	private static String runtime_endpoint = System.getenv("QNA_MAKER_RUNTIME_ENDPOINT");
	// </resourceKeys>

	// <authenticate>
	/* Note QnAMakerManager.authenticate() does not set the baseUrl paramater value
	 * as the value for QnAMakerClient.endpoint, so we still need to call withEndpoint().
	 */
	QnAMakerClient authoring_client = QnAMakerManager.authenticate(authoring_key).withEndpoint(authoring_endpoint);
	Knowledgebases kb_client = authoring_client.knowledgebases();
	Operations ops_client = authoring_client.operations();
	EndpointKeys keys_client = authoring_client.endpointKeys();
	// </authenticate>

	// <listKbs>
	public void list_kbs() {
		System.out.println("Listing KBs...");
		var result = kb_client.listAllAsync();
		result.subscribe(kbs -> {
			for (var kb : kbs.knowledgebases()) {
				System.out.print(kb.id().toString());
			};
		});
		System.out.println();
	}
	// </listKbs>

	// <waitForOperation>
	public String wait_for_operation(Operation op) throws Exception {
		System.out.println ("Waiting for operation to finish...");
		Boolean waiting = true;
		String result = "";
		while (true == waiting) {
			var op_ = ops_client.getDetails(op.operationId());
			var state = op_.operationState();
			if (OperationStateType.FAILED == state) {
				throw new Exception("Operation failed.");
			}
			if (OperationStateType.SUCCEEDED == state) {
				waiting = false;
				// Remove "/knowledgebases/" from the resource location.
				result = op_.resourceLocation().replace("/knowledgebases/", "");
			}
			if (true == waiting) {
				System.out.println("Waiting 10 seconds for operation to complete...");
				Thread.sleep(10000);
			}
		}
		return result;
	}
	// </waitForOperation>

	// <createKb>
	public String create_kb () throws Exception {
		System.out.println("Creating KB...");

		String name = "QnA Maker FAQ from quickstart";

		var metadata = new MetadataDTO()
			.withName ("Category")
			.withValue ("api");

		List<MetadataDTO> metadata_list = Arrays.asList(new MetadataDTO[]{ metadata });

		var qna = new QnADTO()
			.withAnswer ("You can use our REST APIs to manage your knowledge base.")
			.withQuestions ( Arrays.asList(new String[]{ "How do I manage my knowledgebase?" }))
			.withMetadata (metadata_list);

		List<QnADTO> qna_list = Arrays.asList(new QnADTO[]{ qna });

		var urls = Arrays.asList(new String[]{ "https://docs.microsoft.com/en-in/azure/cognitive-services/qnamaker/faqs" });

		var payload = new CreateKbDTO().withName(name).withQnaList(qna_list).withUrls(urls);

		var result = kb_client.create(payload);
		var kb_id = wait_for_operation(result);

		System.out.println("Created KB with ID: " + kb_id + ".\n");
		return kb_id;
	}
	// </createKb>

	// <publishKb>
	public void publish_kb(String kb_id) {
		System.out.println("Publishing KB...");
		kb_client.publish(kb_id);
		System.out.println("KB published.\n");
	}
	// </publishKb>

	// <downloadKb>
	public void download_kb(String kb_id) {
		System.out.println("Downloading KB...");

		var kb_data = kb_client.download(kb_id, EnvironmentType.PROD);
		System.out.println("KB Downloaded. It has " + kb_data.qnaDocuments().size() + " question/answer sets.");

		System.out.println("Downloaded KB.\n");
	}
	// </downloadKb>

	// <queryKb>
	public void query_kb(String kb_id) {
		System.out.println("Sending query to KB...");
		
		var runtime_key = keys_client.getKeys().primaryEndpointKey();
		QnAMakerRuntimeClient runtime_client = QnAMakerRuntimeManager.authenticate(runtime_key).withRuntimeEndpoint(runtime_endpoint);
		var query = (new QueryDTO()).withQuestion("How do I manage my knowledgebase?");
		var result = runtime_client.runtimes().generateAnswer(kb_id, query);
		System.out.println("Answers:");
		for (var answer : result.answers()) {
			System.out.println(answer.answer().toString());
		};
		System.out.println();
	}
	// </queryKb>

	// <deleteKb>
	public void delete_kb(String kb_id) {
		System.out.println("Deleting KB...");
		kb_client.delete(kb_id);
		System.out.println("KB deleted.\n");
	}
	// </deleteKb>

	// <main>
    public static void main(String[] args) {
		try {
			Quickstart quickstart = new Quickstart();
			String kb_id = quickstart.create_kb();
//			quickstart.list_kbs();
			quickstart.publish_kb(kb_id);
			quickstart.download_kb(kb_id);
			quickstart.query_kb(kb_id);
			quickstart.delete_kb(kb_id);
        } catch (Exception e) {
            System.out.println(e.getMessage());
            e.printStackTrace();
        }
    }
	// </main>
}